### PR TITLE
Add missing break and fallthrough comment within switch-case

### PR DESCRIPTION
**Pull request by gfphoenix78**
GCC 7+ has a compiling option, -Wimplicit-fallthrough, which will
generate a warning/error if the code falls through cases(or default)
implicitly. The implicity may cause some bugs that are hardly to catch.

1. Append a comment line /* fallthrough */ or like at the end of case block.
2. Add break clause at the end of case block, if the last statement is
  ereport(ERROR) or like.

This PR is related to #8798 , but without changing configure.in and configure.
Because 5X_STABLE is compiled with a lower version of GCC which does not
support `-Wimplicit-fallthrough`

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/utils/resscheduler/resqueue.c [hunk 1]
```diff
@@ -1324,6 +1324,7 @@ ResCheckSelfDeadLock(LOCK *lock, PROCLOCK *proclock, ResPortalIncrement *increme
 						costThesholdOvercommitted = true;
 					}
 				}
+				break;
```
Comment (gfphoenix78) [inline src/backend/utils/resscheduler/resqueue.c]: This line should fix a bug. Please review.

File: src/backend/executor/nodeAgg.c [hunk 2]
```diff
@@ -1057,6 +1057,7 @@ ExecAgg(AggState *node)
 					 * pass through. Be sure that the next case statement
 					 * is HASHAGG_END_OF_PASSES.
 					 */
+					/* fallthrough */
```
Comment (asimrp) [inline src/backend/executor/nodeAgg.c]: Redundant, the comment above already indicates it's a pass through.

File: src/backend/regex/regc_lex.c [hunk 3]
```diff
@@ -854,6 +854,7 @@ lexescape(struct vars * v)
 			/* oops, doesn't look like it's a backref after all... */
 			v->now = save;
 			/* and fall through into octal number */
+			/* fallthrough */
```
Comment (asimrp) [inline src/backend/regex/regc_lex.c]: Redundant

File: src/backend/regex/regcomp.c [hunk 4]
```diff
@@ -863,6 +863,7 @@ parseqatom(struct vars * v,
 			/* legal in EREs due to specification botch */
 			NOTE(REG_UPBOTCH);
 			/* fallthrough into case PLAIN */
+			/* fallthrough */
```
Comment (asimrp) [inline src/backend/regex/regcomp.c]: Redundant

File: src/backend/utils/adt/datetime.c [hunk 5]
```diff
@@ -3448,6 +3448,7 @@ DecodeISO8601Interval(char *str,
 						continue;
 					}
 					/* Else fall through to extended alternative format */
+					/* FALL THROUGH */
```
Comment (asimrp) [inline src/backend/utils/adt/datetime.c]: Redundant

File: src/backend/utils/misc/faultinjector.c [hunk 6]
```diff
@@ -1577,6 +1577,7 @@ FaultInjector_IsFaultInjected(
 			
 			retval = TRUE;
 			/* NO break */
+			/* fallthrough */
```
Comment (asimrp) [inline src/backend/utils/misc/faultinjector.c]: Redundant

File: src/backend/utils/resscheduler/resqueue.c (hunk 1 above)
Comment (asimrp) [reply src/backend/utils/resscheduler/resqueue.c]: If the `break` is missing, `memoryThesholdOvercommitted` is also set when `costThesholdOvercommitted` is set.  Looking at the remaining part of this function, it doesn't seem to be a serious problem.  But if it is, can we have a test case to demonstrate the failure?

File: src/backend/utils/resscheduler/resscheduler.c [hunk 7]
```diff
@@ -583,6 +583,7 @@ ResLockPortal(Portal portal, QueryDesc *qDesc)
 					break;
 				}
 			}
+			/* fallthrough */
```
Comment (asimrp) [inline src/backend/utils/resscheduler/resscheduler.c]: Redundant.

File: src/backend/utils/sort/tuplestore.c [hunk 8]
```diff
@@ -868,6 +868,7 @@ tuplestore_gettuple(Tuplestorestate *state, bool forward,
 					elog(ERROR, "tuplestore seek failed");
 			state->status = TSS_READFILE;
 			/* FALL THRU into READFILE case */
+			/* fallthrough */
```
Comment (asimrp) [inline src/backend/utils/sort/tuplestore.c]: Redundant

File: src/interfaces/ecpg/pgtypeslib/interval.c [hunk 9]
```diff
@@ -200,6 +200,7 @@ DecodeISO8601Interval(char *str,
 						continue;
 					}
 					/* Else fall through to extended alternative format */
+					/* FALL THROUGH */
```
Comment (asimrp) [inline src/interfaces/ecpg/pgtypeslib/interval.c]: Redundant

File: src/backend/cdb/cdbgroup.c [hunk 10]
```diff
@@ -2205,35 +2205,35 @@ make_plan_for_one_dqa(PlannerInfo *root, MppGroupContext *ctx, int dqa_index,
...
-		groups_sorted = false;
-		break;
-		
-	case DQACOPLAN_PGS:
-	case DQACOPLAN_PH:
-		/* plainagg */
-		aggstrategy = AGG_PLAIN;
-		groups_sorted = false;
-		break;
+		case DQACOPLAN_GSH:
```
Comment (asimrp) [inline src/backend/cdb/cdbgroup.c]: Whitespace changes should be in a separate commit for ease of review.  That said, let's avoid making any whitespace changes on a stable branch.

Comment (gfphoenix78) [reply src/backend/cdb/cdbgroup.c]: OK, I'll take this rule.

**Comment by gfphoenix78** [comment]
@asimrp Redundant is not redundant.
With GCC >= 7, the fallthrough comment is not just a comment, it tells the compiler not to generate implicit-fallthrough warning/error. The old comment about fallthrough can't recognized by GCC, but the added comment can be recognized.
If someone wants to compile gpdb with `-Wimplicit-fallthrough` on, the added comment will help, then only a few places need to check.

**Comment by pivotal-ning-yu** [comment]
FYI, here is the comment patterns accepted by gcc (copied from gcc 9.2.1's manpage):

           C++17 provides a standard way to suppress the -Wimplicit-fallthrough warning using
           "[[fallthrough]];" instead of the GNU attribute.  In C++11 or C++14 users can use
           "[[gnu::fallthrough]];", which is a GNU extension.  Instead of these attributes, it is
           also possible to add a fallthrough comment to silence the warning.  The whole body of
           the C or C++ style comment should match the given regular expressions listed below.
           The option argument n specifies what kind of comments are accepted:

           *<-Wimplicit-fallthrough=0 disables the warning altogether.>
           *<-Wimplicit-fallthrough=1 matches ".*" regular>
               expression, any comment is used as fallthrough comment.

           *<-Wimplicit-fallthrough=2 case insensitively matches>
               ".*falls?[ \t-]*thr(ough|u).*" regular expression.

           *<-Wimplicit-fallthrough=3 case sensitively matches one of the>
               following regular expressions:

               *<"-fallthrough">
               *<"@fallthrough@">
               *<"lint -fallthrough[ \t]*">
               *<"[ \t.!]*(ELSE,? |INTENTIONAL(LY)? )?FALL(S | |-)?THR(OUGH|U)[
               \t.!]*(-[^\n\r]*)?">
               *<"[ \t.!]*(Else,? |Intentional(ly)? )?Fall((s | |-)[Tt]|t)hr(ough|u)[
               \t.!]*(-[^\n\r]*)?">
               *<"[ \t.!]*([Ee]lse,? |[Ii]ntentional(ly)? )?fall(s | |-)?thr(ough|u)[
               \t.!]*(-[^\n\r]*)?">
           *<-Wimplicit-fallthrough=4 case sensitively matches one of the>
               following regular expressions:

               *<"-fallthrough">
               *<"@fallthrough@">
               *<"lint -fallthrough[ \t]*">
               *<"[ \t]*FALLTHR(OUGH|U)[ \t]*">
           *<-Wimplicit-fallthrough=5 doesn't recognize any comments as>
               fallthrough comments, only attributes disable the warning.

           The comment needs to be followed after optional whitespace and other comments by "case"
           or "default" keywords or by a user label that precedes some "case" or "default" label.

                   switch (cond)
                     {
                     case 1:
                       bar (0);
                       /* FALLTHRU */
                     default:
                       ...
                     }

File: src/backend/utils/resscheduler/resqueue.c (hunk 1 above)
Comment (gfphoenix78) [reply src/backend/utils/resscheduler/resqueue.c]: > If the `break` is missing, `memoryThesholdOvercommitted` is also set when `costThesholdOvercommitted` is set. Looking at the remaining part of this function, it doesn't seem to be a serious problem. But if it is, can we have a test case to demonstrate the failure?

Yes, break or no break, it generates the same result. This is not a bug.
The two cases(`RES_COST_LIMIT` and `RES_MEMORY_LIMIT`) can be merged into one.

Thank you for your review.

File: src/backend/executor/nodeAgg.c (hunk 2 above)
Comment (gfphoenix78) [reply src/backend/executor/nodeAgg.c]: Yes, 5X_STABLE of gpdb is compiled (in pipeline) with a lower version of gcc which doesn't support `-Wimplicit-fallthrough`. So it takes no effect if we enable the compile option for 5X_STABLE.

**Comment by gfphoenix78** [comment]
Thank you, Ning.

**Comment by gfphoenix78** [comment]
> Do we not want to enable the `-Wimplicit-fallthrough` compiler option as well?
> 
> Many of the comments seemed redundant, I've flagged some but must have missed some too.

Yes, gpdb 5X is compiled by gcc with a lower version(less than 7). Enabling this compiler option has no effect in pipeline.
